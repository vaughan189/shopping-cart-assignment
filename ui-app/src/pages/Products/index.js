/* eslint-disable react-hooks/exhaustive-deps */
import React, { useEffect, useState } from "react";
import {
  Box,
  Container,
  FormControl,
  Grid,
  MenuItem,
  Select,
} from "@mui/material";

import { Sidebar } from "../../components/Sidebar";
import useAxios from "../../api/useAxios";
import { ProductList } from "./components/ProductList";
import useCart from "../../hooks/useCart";

export const Products = () => {
  const [filterSelection, setFilterSelection] = useState("");
  const [productsList, setProductsList] = useState([]);
  const [filteredProductsList, setFilteredProductsList] = useState([]);

  const { addProductsToCart } = useCart();

  const {
    response: categoriesData,
    loading: categoriesDataLoader,
    fetchData: fetchCategoriesData,
  } = useAxios();

  const {
    response: productsData,
    loading: productsDataLoader,
    fetchData: fetchProductsData,
  } = useAxios();

  const getProductsList = () => {
    fetchProductsData({
      method: "GET",
      url: "http://localhost:3000/products",
      headers: {
        accept: "*/*",
      },
    });
  };

  const getCategoriesList = () => {
    fetchCategoriesData({
      method: "GET",
      url: "http://localhost:3000/categories",
      headers: {
        accept: "*/*",
      },
    });
  };

  const handleCategorySelection = (id) => {
    if (filterSelection !== id) setFilterSelection(id);
    else if (filterSelection === id) setFilterSelection("");
  };

  useEffect(() => {
    getProductsList();
    getCategoriesList();
  }, []);

  useEffect(() => {
    if (
      !productsDataLoader &&
      Array.isArray(productsData.data) &&
      productsData.data.length > 0
    ) {
      setProductsList(productsData.data);
      setFilteredProductsList(productsData.data);
    }
  }, [productsData]);

  useEffect(() => {
    if (filterSelection)
      setFilteredProductsList(
        productsList.filter((el) => el.category === filterSelection)
      );
    else setFilteredProductsList(productsList);
  }, [filterSelection]);

  return (
    <Box component="main" sx={{ marginTop: "2%" }}>
      <Container maxWidth="xl">
        <Grid container direction="row" justifyContent="flex-start">
          <Box
            sx={{
              marginTop: 2,
              marginBottom: 2,
              display: { xs: "contents", sm: "none" },
            }}
          >
            <FormControl fullWidth>
              <Select
                id="demo-simple-select"
                sx={{ color: "#FFFFFF", backgroundColor: "#c73f6d" }}
                value={filterSelection}
                MenuProps={{}}
              >
                <MenuItem value="" onClick={() => handleCategorySelection("")}>
                  Select Category
                </MenuItem>
                {categoriesData &&
                  categoriesData.data &&
                  categoriesData.data.map((item) => {
                    return (
                      <MenuItem
                        key={item.key}
                        value={item.id}
                        onClick={() => handleCategorySelection(item.id)}
                      >
                        {item.name}
                      </MenuItem>
                    );
                  })}
              </Select>
            </FormControl>
          </Box>
          <Box
            sx={{
              display: { xs: "none", sm: "contents" },
            }}
          >
            <Grid item xs={2} sm={2} md={2} xl={2}>
              {!categoriesDataLoader && (
                <Grid container direction="column">
                  <Sidebar
                    categoriesData={categoriesData.data}
                    handleCategorySelection={handleCategorySelection}
                  />
                </Grid>
              )}
            </Grid>
          </Box>
          <Grid item xs={12} sm={10} md={10} xl={10}>
            {!productsDataLoader && (
              <Grid container direction="column">
                <ProductList
                  productsData={filteredProductsList}
                  addProductsToCart={addProductsToCart}
                />
              </Grid>
            )}
          </Grid>
        </Grid>
      </Container>
    </Box>
  );
};
