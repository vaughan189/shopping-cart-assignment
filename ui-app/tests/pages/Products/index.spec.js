import React from "react";
import { shallow } from "enzyme";
import Products from "../../../src/pages/Products";

jest.mock("react-router-dom", () => ({
  useHistory: () => ({
    push: jest.fn(),
  }),
}));

describe("Products Page", () => {
  it("should render the Products Page correctly", () => {
    const component = shallow(<Products debug />);
    expect(component).toMatchSnapshot();
  });
});
