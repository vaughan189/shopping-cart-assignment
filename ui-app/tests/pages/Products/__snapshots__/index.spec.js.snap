// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Products Page should render the Products Page correctly 1`] = `
<ForwardRef(Typography)
  component="div"
  sx={
    Object {
      "marginTop": "1%",
    }
  }
>
  <ForwardRef(Container)
    maxWidth="xl"
  >
    <ForwardRef(Grid)
      container={true}
      direction="row"
      justifyContent="flex-start"
    >
      <ForwardRef(Typography)
        component="div"
        sx={
          Object {
            "display": Object {
              "sm": "none",
              "xs": "contents",
            },
            "marginBottom": 2,
            "marginTop": 2,
          }
        }
      >
        <ForwardRef(FormControl)
          fullWidth={true}
        >
          <ForwardRef(InputLabel)
            htmlFor="categories-menu"
          >
            Select Category
          </ForwardRef(InputLabel)>
          <ForwardRef(Select)
            MenuProps={Object {}}
            id="categories-menu"
            sx={
              Object {
                "backgroundColor": "#c73f6d",
                "color": "#FFFFFF",
              }
            }
            value=""
          >
            <ForwardRef(MenuItem)
              onClick={[Function]}
              value=""
            >
              Select Category
            </ForwardRef(MenuItem)>
          </ForwardRef(Select)>
        </ForwardRef(FormControl)>
      </ForwardRef(Typography)>
      <ForwardRef(Typography)
        component="div"
        sx={
          Object {
            "display": Object {
              "sm": "contents",
              "xs": "none",
            },
          }
        }
      >
        <ForwardRef(Grid)
          item={true}
          md={2}
          sm={2}
          xl={2}
          xs="auto"
        />
      </ForwardRef(Typography)>
      <ForwardRef(Grid)
        item={true}
        md={10}
        sm={10}
        xl={10}
        xs={12}
      />
    </ForwardRef(Grid)>
  </ForwardRef(Container)>
</ForwardRef(Typography)>
`;
